sycl/include/CL/sycl/ONEAPI/group_algorithm.hpp

```diff
@@ -359,10 +197,13 @@ broadcast(Group g, T x, typename Group::linear_id_type linear_local_id) {
 }
 
 template <typename Group, typename T>
+__SYCL2020_DEPRECATED(
+    "ONEAPI::broadcast is deprecated. Use group_broadcast instead.")
```

Comment:
Why don't we use `group_broadcast` to implement `ONEAPI::broadcast`?
The question applies to `reduce`.

Reply:
In one of the earlier iterations of this PR, that's exactly what I did. But there is a signature difference involving `is_trivially_copyable` that makes them (SYCL2020 and ONEAPI) incompatible.  The SYCL2020 version should not provide functionality that exceeds the specification, in case we need to later take that away.     The discussion is on this PR somewhere - I'll see if I can find it.

Reply:
If ONEAPI and SYCL 2020 version are not exactly the same, does it make sense to deprecate ONEAPI version? It sounds like ONEAPI version "provides functionality that exceeds the specification".

Reply:
This is true but expected to be temporary.  The group algorithms in `sycl::ONEAPI::` are extensions against SYCL 1.2.1 that don't reflect the names or other conventions of the group algorithms that eventually made it into SYCL 2020, and that could confuse developers.

I've got an issue open at Khronos to discuss whether SYCL 2020 broadcast should work with all trivially copyable types or if it should be restricted only to fundamental types.  If the decision comes back that SYCL 2020 broadcast should only support fundamental types and we want to lift that restriction in our implementation, we should write a new extension against SYCL 2020.  We need to write a new extension to use SYCL 2020 sub-groups, adopt consistent naming (e.g. `reduce_over_group` and `joint_reduce` vs `reduce`), and obey the SYCL 2020 extension mechanism (i.e. we need to use the `sycl::ext::oneapi` namespace).

Reply:
I think it makes sense.  Most users are not going to wander into the crevices between these two APIs, that's unlikely. We fully intend to deprecate the ONEAPI version at some point in the future.  So we should definitely emit the deprecation warning. Otherwise, we'd have to commit to keeping the ONEAPI:broadcast version around in the future and supporting it.   

Here are the previous threads on this: https://github.com/intel/llvm/pull/3786#discussion_r636127542
and https://github.com/intel/llvm/pull/3786#discussion_r636132468

Reply:
> In one of the earlier iterations of this PR, that's exactly what I did. But there is a signature difference involving `is_trivially_copyable` that makes them (SYCL2020 and ONEAPI) incompatible. The SYCL2020 version should not provide functionality that exceeds the specification, in case we need to later take that away. The discussion is on this PR somewhere - I'll see if I can find it.

Do you refer this this comment: https://github.com/intel/llvm/pull/3786#discussion_r636132468?

Reply:
Thanks for clarification. I've got that we want to deprecate existing extension anyway.